infrastructure/comms/src/connection/net_address.rs

```diff
@@ -0,0 +1,91 @@
+//  Copyright 2018 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use std::str::FromStr;
+
+use connection::{i2p::I2PAddress, onion::OnionAddress, p2p::SocketAddress};
+
+#[derive(Debug, Error)]
+pub enum NetAddressError {
+    /// Failed to parse address
+    ParseFailed,
+    /// Specified port range is invalid
+    InvalidPortRange,
+}
+
+/// Represents an address which can be used to reach a node on the network
+pub enum NetAddress {
+    /// IPv4 and IPv6
+    IP(SocketAddress),
+    Tor(OnionAddress),
+    I2P(I2PAddress),
+}
+
+impl FromStr for NetAddress {
+    type Err = NetAddressError;
+
+    fn from_str(address: &str) -> Result<Self, Self::Err> {
```

Comment:
Would it not be better to have a function interpret the string and then do this via a match?

I dont know how rust deals with fails eg (errors) but other languages are really slow if you rely on the "catch" part to function. In effect here we try to parse, them, if they fail we try the next one. If we do this the order is going to very important here as well.

Reply:
Agree that this is brittle - as it matters which order they are parsed in. The slowness would come from running each parse function as this isn't really at all like multiple try catches in other languages. Agree `FromStr` may not be a good fit for `NetAddress` ref https://github.com/tari-project/tari/issues/146